### ISSUE TITLE
Testes Unitários no modelo Qwen 3

### ISSUE THREAD

**Issue by diegomrodrigues2** [diegomrodrigues2/py_distributed_database#172]
### 1. Validação de Shape das Saídas

Crie inputs sintéticos, por exemplo `x = torch.randn(batch_size, seq_len, hidden_size)`, e aplique diretamente módulos isolados do Qwen3 (como `Qwen3Attention`, `Qwen3MLP` ou `RMSNorm`), verificando que `output.shape == (batch_size, seq_len, hidden_size)` ou o formato esperado para cada camada ([[stackoverflow.com](https://stackoverflow.com/questions/63917991/unittest-the-pytorch-forward-function?utm_source=chatgpt.com)][1], [[stackoverflow.com](https://stackoverflow.com/questions/55875279/how-to-get-an-output-dimension-for-each-layer-of-the-neural-network-in-pytorch?utm_source=chatgpt.com)][2]).

### 2. Máscara de Atenção (Attention Mask)

Construa uma máscara `attention_mask` onde algumas posições são zero e passe-a a um `TransformerEncoderLayer`. Assegure que, antes do softmax, os scores correspondentes às posições mascaradas sejam `-inf`, por exemplo com:

```python
scores = layer.self_attn(query, key, value, attn_mask=mask)[1]
assert torch.all(scores[mask == 0] == float('-inf'))
```

([[stackoverflow.com](https://stackoverflow.com/questions/76034630/how-do-i-send-an-attention-mask-mask-matrix-in-transformer-encoder-along-with?utm_source=chatgpt.com)][3])

### 3. Teste de RMSNorm

Alimente o módulo de normalização RMSNorm com tensores de média e variância conhecidas e confirme que a saída possui média ≈ 0 e variância ≈ 1 em cada canal:

```python
y = rmsnorm(x)
assert torch.allclose(y.mean(dim=-1), 0, atol=1e-5)
assert torch.allclose(y.var(dim=-1), 1, atol=1e-5)
```

([[hypothesis-torch.readthedocs.io](https://hypothesis-torch.readthedocs.io/en/stable/quick_start_guide/?utm_source=chatgpt.com)][4])

### 4. Herança e Atributos da Configuração

Instancie `cfg = Qwen3Config()` e verifique:

* `isinstance(cfg, PretrainedConfig)`
* Valores padrão de `cfg.n_layer`, `cfg.n_head`, `cfg.hidden_size` correspondem à documentação do Qwen3 ([[huggingface.co](https://huggingface.co/docs/transformers/testing?utm_source=chatgpt.com)][5]).

### 5. Embeddings com IDs Fixos

Forneça IDs fixos, como `ids = torch.tensor([[1, 2, 3]])`, e compare diretamente:

```python
emb = model.get_input_embeddings().weight
assert torch.equal(emb[ids], model.get_input_embeddings()(ids))
```

Isso garante que mudanças não alterem os índices de embedding ([[github.com](https://github.com/suriyadeepan/torchtest?utm_source=chatgpt.com)][6]).

### 6. Camada MLP Isolada

Extraia o sub-módulo MLP (por exemplo `model.qwen3_mlp`) e alimente um tensor `x`. Compare a saída com o cálculo manual de `x @ W1 + b1` seguido de `GELU` e `x @ W2 + b2`:

```python
manual = F.gelu(x @ W1 + b1) @ W2 + b2
assert torch.allclose(mlp(x), manual, atol=1e-6)
```

([[github.com](https://github.com/suriyadeepan/torchtest?utm_source=chatgpt.com)][6])

---

### 1. Salvamento e Recarga de Pesos (`state_dict`)

Treine ou ajuste uma instância dummy do modelo, faça:

```python
torch.save(model.state_dict(), "tmp.pth")
model2 = Qwen3Model(cfg)
model2.load_state_dict(torch.load("tmp.pth"))
```

E então compare cada tensor de `model.parameters()` com `model2.parameters()` usando `torch.equal` ([[github.com](https://github.com/huggingface/transformers/issues/14642?utm_source=chatgpt.com)][7]).

### 2. Chamada Direta da API de Forward

Use `outputs = model(**inputs)`, onde `inputs` vem de `tokenizer` (com `input_ids` e `attention_mask`). Verifique que:

* `“logits” in outputs.keys()`
* `outputs.logits.shape == (batch_size, seq_len, vocab_size)`
* `“past_key_values” in outputs.keys()`
  Tudo sem invocar `generate()` ([[discuss.huggingface.co](https://discuss.huggingface.co/t/using-model-instead-of-model-generate/116329?utm_source=chatgpt.com)][8]).

### 3. Compatibilidade com DataParallel

Envolva o modelo em `torch.nn.DataParallel(model)` e compare:

```python
out_dp = dp_model(**inputs)
out_orig = model(**inputs)
assert torch.allclose(out_dp.logits, out_orig.logits, atol=1e-6)
```

Isso garante que replicação multilâmina não altera resultados ([[github.com](https://github.com/pytorch/pytorch/wiki/Running-and-writing-tests?utm_source=chatgpt.com)][9]).

### 4. Integração com Tokenizer

Combine `Qwen3Tokenizer` e `Qwen3Model` num fluxo único:

```python
tok_out = tokenizer("Olá mundo", return_tensors="pt")
model(**tok_out)
```

Assegure que não ocorram erros de shape ou tipos, validando interoperabilidade entre tokenização e modelo ([[huggingface.co](https://huggingface.co/docs/transformers/testing?utm_source=chatgpt.com)][5]).

### 5. Weight Tying Consistente

Verifique que o embedding de entrada e o head de saída compartilham parâmetros:

```python
assert model.get_input_embeddings().weight.data_ptr() == model.get_output_embeddings().weight.data_ptr()
```

Garantindo que alterações não quebrem o weight tying ([[github.com](https://github.com/huggingface/transformers/issues/14642?utm_source=chatgpt.com)][7]).

[1]: https://stackoverflow.com/questions/63917991/unittest-the-pytorch-forward-function?utm_source=chatgpt.com "python - Unittest the pytorch forward function - Stack Overflow"
[2]: https://stackoverflow.com/questions/55875279/how-to-get-an-output-dimension-for-each-layer-of-the-neural-network-in-pytorch?utm_source=chatgpt.com "How to get an output dimension for each layer of the Neural Network in ..."
[3]: https://stackoverflow.com/questions/76034630/how-do-i-send-an-attention-mask-mask-matrix-in-transformer-encoder-along-with?utm_source=chatgpt.com "python - How do I send an attention-mask \"Mask\" matrix in transformer ..."
[4]: https://hypothesis-torch.readthedocs.io/en/stable/quick_start_guide/?utm_source=chatgpt.com "Quick Start Guide - hypothesis-torch"
[5]: https://huggingface.co/docs/transformers/testing?utm_source=chatgpt.com "Testing - Hugging Face"
[6]: https://github.com/suriyadeepan/torchtest?utm_source=chatgpt.com "GitHub - suriyadeepan/torchtest: Unit Testing for pytorch, based on mltest"
[7]: https://github.com/huggingface/transformers/issues/14642?utm_source=chatgpt.com "Encapsulate all forward passes of integration tests with \"with ... - GitHub"
[8]: https://discuss.huggingface.co/t/using-model-instead-of-model-generate/116329?utm_source=chatgpt.com "Using model() instead of model.generate() - Transformers - Hugging ..."
[9]: https://github.com/pytorch/pytorch/wiki/Running-and-writing-tests?utm_source=chatgpt.com "Running and writing tests · pytorch/pytorch Wiki · GitHub"
